extensions/smallrye-jwt/deployment/src/main/java/io/quarkus/smallrye/jwt/deployment/SmallRyeJwtProcessor.java

```diff
@@ -161,14 +165,35 @@ void registerOptionalClaimProducer(BeanRegistrationPhaseBuildItem beanRegistrati
                 continue;
             }
             AnnotationInstance claimQualifier = injectionPoint.getRequiredQualifier(CLAIM_NAME);
-            if (claimQualifier != null && injectionPoint.getType().name().equals(DotNames.PROVIDER)) {
-                // Classes from jakarta.json are handled specially
+            if (claimQualifier != null) {
+                boolean providerType = injectionPoint.getType().name().equals(DotNames.PROVIDER);
+                boolean instanceType = injectionPoint.getType().name().equals(DotNames.INSTANCE);
+
                 Type actualType = injectionPoint.getRequiredType();
-                if (actualType.name().equals(DotNames.OPTIONAL) && !actualType.name().toString()
-                        .startsWith("jakarta.json")) {
-                    additionalTypes.add(actualType);
+
+                Optional<BeanInfo> bean = injectionPoint.getTargetBean();
+                if (bean.isPresent()) {
+                    DotName scope = bean.get().getScope().getDotName();
+                    if ((DotNames.SINGLETON.equals(scope) || DotNames.APPLICATION_SCOPED.equals(scope))
```

Comment:
What happens if scope is `Dependent` and the bean is injected into `Singleton` bean?

Reply:
I mean, if the claim is dependent injected into dependent bean injected into singleton, it will still have stale value, will it not?

Reply:
@michalvavrik I see, so I believe I need to tune and fail the build if the raw @Claim is injected in any but `@RequestScoped` bean, correct ?

Reply:
I think it would be the simplest way yes. You can also keep iterate over bean injection tree until you determine non-dependent one. Personally I'd require `@RequestScope`.

Reply:
Indeed, just requiring `RequestScoped` simplifies, thanks